File: src/test/java/travelator/money/ExchangeRatesViaBaseCurrencyTest.java
package travelator.money;

import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.math.BigDecimal;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static travelator.money.Currencies.*;

public class ExchangeRatesViaBaseCurrencyTest {
    private final ExchangeRates fx = new ExchangeRatesViaBaseCurrency(
        EUR,
        Map.entry(GBP, new BigDecimal("0.8")),
        Map.entry(JOD, new BigDecimal("100"))
    );

    @Test
    public void same_currency() throws IOException {
        assertEquals(
            Money.of("125", EUR),
            fx.convert(Money.of("125", EUR), EUR).getToMoney());
    }

    @Test
    public void one_foreign_currency() throws IOException {
        assertConversionBothWays(Money.of("100", EUR), Money.of("80", GBP));
    }

    @Test
    public void two_foreign_currencies() throws IOException {
        assertConversionBothWays(Money.of("80", GBP), Money.of("10000", JOD));
    }

    public void assertConversionBothWays(Money m1, Money m2) throws IOException {
        assertConversion(m1, m2);
        assertConversion(m2, m1);
    }

    public void assertConversion(Money fromMoney, Money toMoney) throws IOException {
        assertEquals(
            toMoney,
            fx.convert(fromMoney, toMoney.getCurrency()).getToMoney(),
            () -> fromMoney.getCurrency() + " -> " + toMoney.getCurrency());
    }
}

File: src/test/java/travelator/money/ExchangeRatesViaBaseCurrencyTest.kt
package travelator.money

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import travelator.money.Money.Companion.of
import java.math.BigDecimal

class ExchangeRatesViaBaseCurrencyTest {
    private val fx: ExchangeRates = ExchangeRatesViaBaseCurrency(
        EUR,
        mapOf(
            GBP to BigDecimal("0.8"),
            JOD to BigDecimal("100")))

    @Test
    fun same_currency() {
        assertEquals(
            Money("125", EUR),
            fx.convert(of("125", EUR), EUR).toMoney)
    }

    @Test
    fun one_foreign_currency() {
        assertConversionBothWays(of("100", EUR), Money("80", GBP))
    }

    @Test
    fun two_foreign_currencies() {
        assertConversionBothWays(of("80", GBP), Money("10000", JOD))
    }

    private fun assertConversionBothWays(m1: Money, m2: Money) {
        assertConversion(m1, m2)
        assertConversion(m2, m1)
    }

    private fun assertConversion(fromMoney: Money, toMoney: Money) {
        assertEquals(toMoney, fx.convert(fromMoney, toMoney.currency).toMoney) {
            fromMoney.currency.toString() + " -> " + toMoney.currency
        }
    }
}